### ISSUE TITLE
System defaults not applied properly nor globally

### ISSUE THREAD

**Issue by JanKleine** [LnL7/nix-darwin#1207]
Hi,

I recently setup a new MacBook with nix-darwin as I wanted to give it a try for a while.

I noticed that the `system.defaults...` don't seem to be set properly nor globally (for all users). It's entirely possible that I made a mistake as I'm very new to this.

I have the following defaults:
```nix
﻿system.defaults = {
    controlcenter.BatteryShowPercentage = true;
    finder.FXPreferredViewStyle = "Nlsv";
    screensaver.askForPassword = true;
    screensaver.askForPasswordDelay = 0;
    trackpad = {
        Clicking = true;
        TrackpadThreeFingerDrag = true;
    };
};
```

Firstly I noticed that `trackpad.Clicking` seems to work[^1], however in System Settings > Trackpad > Tap to click is still disabled. It is technically working I guess, but gives me the slightly uneasy feeling that settings are not being applied properly. The same applies to `trackpad.TrackpadThreeFingerDrag` and System Settings > Accessibility > Pointer Control > Trackpad Options... > Use trackpad for dragging.

`controlcenter.BatteryShowPercentage` on the other hand works just fine and is also reflected in the settings.

`screensaver.askForPassword` and `screensaver.askForPasswordDelay` don't seem to take affect at all, which is more of an issue, especially with these settings being security relevant. This seems to have been mentioned before in #908.

Finally, all these settings do not seem to work for another user on the system. I have the user listed under `users.users`, but even `controlcenter.BatteryShowPercentage`, which is working as expected for the user controlling nix-darwin, is not being applied.

Is this a bug or is there something wrong with my multi user setup? I'm happy to share more details if needed and appreciate any help!

**Basic system infos**:
MacBook Pro with M4 Pro
macOS: 15.1.1
nix-darwin: 6ab87b7c84d4ee873e937108c4ff80c015a40c7a

[^1]: It may be that it only worked after a restart. I don't quite remember, but there was something that only worked after a restart.

**Comment by Enzime** [comment]
Currently `system.defaults` maps one to one with Apple `defaults` however some options in System Settings actually map to multiple `defaults`. We should add a new nix-darwin option like `hardware.trackpad.tapToClick` which sets all the relevant `defaults` for you.

The easiest way to test is usually to run `defaults read > before.txt`, change the setting and then `defaults read > after.txt` and then check what the differences are, sometimes it might change system wide settings as well so you might need to check `sudo defaults read` as well.

We'll need to remove `system.defaults.screensaver.askForPassword*` and add new options to use `sysadminctl` to set them instead, see: https://github.com/mathiasbynens/dotfiles/issues/922#issuecomment-1322698371, similar to: https://github.com/LnL7/nix-darwin/blob/e30a3622b606dffc622305b4bbe1cfe37e78fa40/modules/power/sleep.nix

Currently we only manage defaults for the primary user (the user you run `darwin-rebuild switch` as), we have plans to fix this but we haven't started on it yet see #96

I don't have the time to work on any of the changes I've described, so feel free to make PRs and I'll review them

**Comment by bryanprimus** [comment]
i'm not sure this is the same issue but it is also happened to me

```nix
system.defaults = {
        dock.autohide = true # --> this is working;
	dock.persistent-apps = [
	  "${pkgs.arc-browser}/Applications/Arc.app" # --> this is working;
	];
	NSGlobalDomain."com.apple.swipescrolldirection" = false; # --> this is working;
	NSGlobalDomain.KeyRepeat = 2; # --> this is not working;
	NSGlobalDomain.AppleInterfaceStyle = "Dark"; # --> this is not working;
};

```

**Comment by nalzok** [comment]
For the record, [`system.defaults.trackpad.Clicking`](https://daiderd.com/nix-darwin/manual/index.html#opt-system.defaults.trackpad.Clicking) does not appear to work on macOS 15.2 (24C101). Tapping does not trigger clicking behavior, and "Tap to click" is disabled.

<img width="827" alt="Screenshot 2024-12-26 at 04 23 29" src="https://github.com/user-attachments/assets/c95c6b7e-35ab-4a04-95bc-8ae3a60cbefe" />

Moreover, the option "Look up & data detectors" now defaults to "Force click with One Finger" instead of "Tap with three Fingers", which IIRC is different from previous versions.

**Comment by Samasaur1** [comment]
@bryanprimus Are you sure that setting `AppleInterfaceStyle` does anything? It doesn't appear to when run manually on my Mac; I think it's only useful to read it to check whether or not you're in dark mode. e.g.:
```bash
defaults read -globalDomain AppleInterfaceStyle &> /dev/null && echo "In dark mode" || echo "In light mode"
```

**Comment by thislooksfun** [comment]
I can confirm that setting `AppleInterfaceStyle` does work, just not instantly. `defaults write NSGlobalDomain AppleInterfaceStyle "Dark"` and `defaults delete NSGlobalDomain AppleInterfaceStyle` both change the light/dark mode on next login. I have not yet found a way to apply the change immediately.

**Comment by Samasaur1** [comment]
AppleScript?
```bash
osascript -e 'tell application "System Events" to tell appearance preferences to set dark mode to not dark mode'
```

**Comment by thislooksfun** [comment]
Yeah, AppleScript works, but I was hoping for a way that wouldn't require it and the associated Automation permission.

**Comment by billdaws** [comment]
Not sure if this is fully relevant, but I found this issue when looking for reasons why my setting of `trackpad.Clicking = true;` wasn't working after a rebuild. The observed behavior I saw is similar to what @nalzok observed where the System Settings GUI did not match the actual behavior, and the actual behavior did not match what I configured, e.g. I would rebuild and the "tap to click" functionality would be disabled, despite `Clicking = true`.

I did a `diff` of defaults (and `sudo`) after rebuilding and then after manually toggling in the System Settings GUI and didn't see any significant changes, just some timestamps.

There seems to be a private framework that fixes this issue:
```
/System/Library/PrivateFrameworks/SystemAdministration.framework/Resources/activateSettings -h
```
...by being invoked after rebuilding.

I suppose it could be [called here in defaults-write.nix](https://github.com/nix-darwin/nix-darwin/blob/master/modules/system/defaults-write.nix#L157-L158) to sort of indiscriminately fix problems like this, but I am not really sure what other problems this could create. Thus I figured it's worth mentioning before doing.

**tl;dr** I think this works, any reason to avoid something like this?

```diff
diff --git a/modules/system/defaults-write.nix b/modules/system/defaults-write.nix          
--- a/modules/system/defaults-write.nix                                                     
+++ b/modules/system/defaults-write.nix                                                     
@@ -154,6 +154,10 @@ in                                                                     
         ${optionalString (length dock > 0) ''                                              
           echo >&2 "restarting Dock..."                                                    
           killall -qu ${escapeShellArg config.system.primaryUser} Dock || true             
         ''}                                                                                
+                                                                                           
+        echo >&2 "activating user defaults..."                                             
+        sudo -u ${user} /System/Library/PrivateFrameworks/SystemAdministration.framework/Resources/activateSettings -u                                                                                          
       '';                                                                                  
                                                                                          
   };
 }
```

**Comment by victorhooi** [comment]
I'm also trying to setup a fresh macOS system with Nix-Darwin, and hitting similar issues around getting trackpad preferences to apply immediately.

I did find a workaround, but wanted to double-check with people here, in whether it should be recommended or not, or if there's a more Nix-ish way to do it?

Here is the relevant part of my flake.nix - note that the `com.apple.mouse.tapBehavior` part is commented out - I'd been randomly trying it before in case it helped, but it seems to be a red herring - it doesn't seem to be needed (@Enzime let me know if this is incorrect, and I should have it as well):
```
+      # Which user "owns" user-scoped system defaults below (and homebrew
+      # later). Required as soon as you use any user-scoped option in
+      # system.defaults — activation errors with a clear message asking
+      # for it if you forget.
+      system.primaryUser = "victorhooi";
+
+      # macOS preferences, managed declaratively. Equivalent to a series
+      # of `defaults write <domain> <key> <value>` calls, but typed and
+      # reproducible. Reverts to system defaults if removed from the
+      # flake (well — see caveat below).
+      system.defaults = {
+        trackpad = {
+          Clicking = true;            # tap-to-click
+          TrackpadRightClick = true;  # two-finger right click
+        };
+
+        NSGlobalDomain = {
+          # The trackpad option above writes to the trackpad pref domain;
+          # this writes the equivalent mirror in NSGlobalDomain. Both are
+          # needed for tap-to-click to actually take effect on macOS — a
+          # known macOS quirk, not a nix-darwin one.
+          #"com.apple.mouse.tapBehavior" = 1; # Correction: This actually isn't needed, AFAIK - you just need to apply the settings.
+
+          # Trackpad tracking speed: 0.0 (slowest) to 3.0 (fastest).
+          "com.apple.trackpad.scaling" = 3.0;
+        };
+
+        dock = {
+          # Swipe down on a dock icon → app expose (shows all windows of
+          # that app).
+          showAppExposeGestureEnabled = true;
+        };
+      };
+
```

I then ran `sudo darwin-rebuild switch` - and the tap-to-click preference was not applied immediately.

However, I then ran:
```
/System/Library/PrivateFrameworks/SystemAdministration.framework/Resources/activateSettings -u
```
and the setting was immediately applied (along with my App Expose gesture).

I got that command from here:
* https://github.com/yannbertrand/macos-defaults/issues/203#issuecomment-4125969063

and it's also referenced in this blog post:
* https://medium.com/@zmre/nix-darwin-quick-tip-activate-your-preferences-f69942a93236

where they suggest adding it in a hook:
```
  system.activationScripts.postUserActivation.text = ''
    # Following line should allow us to avoid a logout/login cycle
    /System/Library/PrivateFrameworks/SystemAdministration.framework/Resources/activateSettings -u
  '';
```

I can also confirm that if I then go into System Settings, Trackpad - it shows "Tap to click" as disabled - even though the actual behaviour is that it's enabled:

<img width="720" height="626" alt="Image" src="https://github.com/user-attachments/assets/1b79db4c-36ac-43b2-b11c-cf42272219dc" />

Firstly - is running `activateSettings -u` the correct way to get the macOS system settings in flake.nix to apply?

Should that be run automatically as part of nix-darwin, or alternatively at least documented somewhere, if users are meant to run it manually?

And in terms of the actual System Settings UI not reflecting this - is there any workaround for that currently?

**Comment by victorhooi** [comment]
Also - some other things I noticed.

Showing battery percentage applied immediately after `sudo darwin-rebuild switch` - I didn't need to run `activateSettings -u`:
```
controlcenter.BatteryShowPercentage = true;
```

And changing 24 hours clock - I'm not sure if I have the syntax wrong:
```
controlcenter.BatteryShowPercentage = true;
```
But this didn't seem to apply, either after `sudo darwin-rebuild switch`, or after running `activateSettings -u`.

**Comment by HuaDeity** [comment]
For "Tap to click", besides `system.defaults.trackpad.Clicking`, you should also enable:
```
defaults -currentHost write NSGlobalDomain com.apple.mouse.tapBehavior -bool true
```
In the same way, `system.defaults.trackpad.TrackpadThreeFingerDrag` is also not enough to enable the setting from `System Settings.app` UI.

Since `currentHost` still not supported by `nix-darwin`(see #1721), I recommend using home-manager for control the `defaults` related things, for "Tap to click" and "Three finger drag", the code would be:
```nix
targets.darwin.defaults = {
  "com.apple.AppleMultitouchTrackpad" = {
    Clicking = true;
    TrackpadThreeFingerDrag = true;
    TrackpadThreeFingerHorizSwipeGesture = 0;
    TrackpadThreeFingerTapGesture = 0;
    TrackpadThreeFingerVertSwipeGesture = 0;
  };
  "com.apple.driver.AppleBluetoothMultitouch.trackpad" = {
    Clicking = true;
    TrackpadThreeFingerDrag = true;
    TrackpadThreeFingerHorizSwipeGesture = 0;
    TrackpadThreeFingerTapGesture = 0;
    TrackpadThreeFingerVertSwipeGesture = 0;
  };
};
targets.darwin.currentHostDefaults = {
  NSGlobalDomain = {
    "com.apple.mouse.tapBehavior" = true;
    "com.apple.trackpad.threeFingerDragGesture" = true;
    "com.apple.trackpad.threeFingerHorizSwipeGesture" = 0;
    "com.apple.trackpad.threeFingerTapGesture" = 0;
    "com.apple.trackpad.threeFingerVertSwipeGesture" = 0;
  };
};
```

**Comment by HuaDeity** [comment]
@victorhooi To enable the 24-hour clock, after running `sudo darwin-rebuild switch`, simply log out and log back in (or restart macOS) to apply the change. Did you perhaps paste the wrong code?
